### PR TITLE
Add TLS-ALPN-01 Challenge Type to ACME

### PR DESCRIPTION
**Pull request by cipherboy**
This adds the last missing (widely supported) challenge type to the PKI ACME instance: TLS-ALPN-01. This should allow a nicer experience for anyone using Caddy, nginx, or similar server which natively supports ACME cert issuance without reloads or modifying their HTTP router to server `.well-known/acme-challenge/` from another location.

### CODE REVIEW

**Review by stevendpclark — COMMENTED** [review]
Looks great so far, I don't have any comments beyond adding a test-case to validate this is works.

File: builtin/logical/pki/acme_challenges.go [hunk 1]
```diff
@@ -188,3 +206,247 @@ func ValidateDNS01Challenge(domain string, token string, thumbprint string, conf
...
+			// > The subjectAlternativeName extension MUST contain a single
+			// > dNSName entry where the value is the domain name being
+			// > validated.
+			//
+			// TODO: this does not validate that there are not other SANs
+			// with unknown (to Go) OIDs.
+			if len(cert.DNSNames) != 1 || len(cert.EmailAddresses) > 0 || len(cert.IPAddresses) > 0 || len(cert.URIs) > 0 {
+				return fmt.Errorf("server under test returned a certificate with too many SANs")
+			}
+			if cert.DNSNames[0] != domain {
```
Comment (cipherboy) [inline builtin/logical/pki/acme_challenges.go:299]: This needs to be case insensitive.

Comment (cipherboy) [reply builtin/logical/pki/acme_challenges.go]: Pushed. :-)

**Review by stevendpclark — APPROVED** [review]

File: builtin/logical/pki/acme_challenges_test.go [hunk 2]
```diff
@@ -219,3 +231,473 @@ func TestAcmeValidateDNS01Challenge(t *testing.T) {
...
+			protocols:    []string{ALPNProtocol},
+			token:        "valid",
+			thumbprint:   "valid",
+			shouldFail:   true,
+		}
+		alpnTestCases = append(alpnTestCases, newTc)
+	}
+
+	{
+		// Test case: root without a leaf
```
Comment (kitography) [inline builtin/logical/pki/acme_challenges_test.go:648]: all of these are self-signed, why should this one fail?

**Review by kitography — APPROVED** [review]
This looks great!

File: builtin/logical/pki/acme_challenges.go [hunk 3]
```diff
@@ -188,3 +215,255 @@ func ValidateDNS01Challenge(domain string, token string, thumbprint string, conf
...
+			// Negotiation (TLS ALPN) Challenge:
+			//
+			// > The subjectAlternativeName extension MUST contain a single
+			// > dNSName entry where the value is the domain name being
+			// > validated.
+			//
+			// TODO: this does not validate that there are not other SANs
+			// with unknown (to Go) OIDs.
+			if len(cert.DNSNames) != 1 || len(cert.EmailAddresses) > 0 || len(cert.IPAddresses) > 0 || len(cert.URIs) > 0 {
+				return fmt.Errorf("server under test returned a certificate with too many SANs")
```
[the remark this answers is not in the document]
Comment (cipherboy) [reply builtin/logical/pki/acme_challenges.go:306]: If there are in either case, it'd be a violation of the spec.

As I was talking in a thread on slack with @stevendpclark and @schultz-is , I'm a little hesitant leaking data about challenge responses in general: in the case of a blatantly wrong validation request (e.g., for a domain not under the requester's control), this could be an attacker-controlled value and result in things like log injections, huge messages, information disclosure, ... &c.

Perhaps we could rephrase this as:

```suggestion
				return fmt.Errorf("server under test returned a certificate with incorrect SANs")
```

and leave it at that?

File: builtin/logical/pki/acme_challenges_test.go (hunk 2 above)
Comment (cipherboy) [reply builtin/logical/pki/acme_challenges_test.go:648]: This is a random root CA, presented as a server certificate. It lacks the extensions required by the challenge RFC (no DNSSans and no acmeIdentifier extension) because it is a root CAs. Root CAs, while yes, self-signed, should not validate and violate other required verification steps.

See https://datatracker.ietf.org/doc/html/rfc8737#section-3:

> The client prepares for validation by constructing a self-signed certificate that MUST contain an acmeIdentifier extension and a subjectAlternativeName extension [RFC5280]. The subjectAlternativeName extension MUST contain a single dNSName entry where the value is the domain name being validated. The acmeIdentifier extension MUST contain the SHA-256 digest [FIPS180-4] of the key authorization [RFC8555] for the challenge. The acmeIdentifier extension MUST be critical so that the certificate isn't inadvertently used by non-ACME software.
